service/dap/server_test.go

```diff
@@ -3036,6 +3046,377 @@ func TestUnupportedCommandResponses(t *testing.T) {
+					tester.expectSetVariable(valRef, "val", "3")
+					tester.evaluate("val", `main.A {val: 3}`, hasChildren)
+				},
+				disconnect: true,
+			}})
+	})
+}
+
+// TestSetVariableWithCall tests SetVariable features that do not depend on function calls support.
+func TestSetVariableWithCall(t *testing.T) {
+	protest.MustSupportFunctionCalls(t, testBackend)
+
+	runTest(t, "testvariables", func(client *daptest.Client, fixture protest.Fixture) {
+		runDebugSessionWithBPs(t, client, "launch",
+			func() {
+				client.LaunchRequestWithArgs(map[string]interface{}{
+					"mode": "exec", "program": fixture.Path, "showGlobalVariables": true,
+				})
+			},
+			fixture.Source, []int{66, 67},
+			[]onBreakpoint{{
+				execute: func() {
+					tester := &helperForSetVariable{t, client}
+
+					startLineno := 66
+					if runtime.GOOS == "windows" && goversion.VersionAfterOrEqual(runtime.Version(), 1, 15) {
+						// Go1.15 on windows inserts a NOP after the call to
+						// runtime.Breakpoint and marks it same line as the
+						// runtime.Breakpoint call, making this flaky, so skip the line check.
+						startLineno = -1
+					}
+
+					handleStop(t, client, 1, "main.foobar", startLineno)
+
+					// Args of foobar(baz string, bar FooBar)
+					args := tester.variables(1000)
+
+					expectVarExact(t, args, 0, "baz", "baz", `"bazburzum"`, noChildren)
+					tester.expectSetVariable(1000, "baz", `"BazBurZum"`)
+					tester.evaluate("baz", `"BazBurZum"`, noChildren)
+
+					args = tester.variables(1000)
+					barRef := expectVarExact(t, args, 1, "bar", "bar", `main.FooBar {Baz: 10, Bur: "lorem"}`, hasChildren)
+					tester.expectSetVariable(barRef, "Bur", `"ipsum"`)
+					tester.evaluate("bar", `main.FooBar {Baz: 10, Bur: "ipsum"}`, hasChildren)
+
+					// Local variables
+					locals := tester.variables(1001)
+
+					expectVarExact(t, locals, -1, "a1", "a1", `"foofoofoofoofoofoo"`, noChildren)
+					tester.expectSetVariable(1001, "a1", `"barbarbar"`)
+					tester.evaluate("a1", `"barbarbar"`, noChildren)
+
+					a6Ref := expectVarExact(t, locals, -1, "a6", "a6", `main.FooBar {Baz: 8, Bur: "word"}`, hasChildren)
+					tester.failSetVariable(a6Ref, "Bur", "false", "can not convert")
+
+					tester.expectSetVariable(a6Ref, "Bur", `"sentence"`)
+					tester.evaluate("a6", `main.FooBar {Baz: 8, Bur: "sentence"}`, hasChildren)
+				},
+			}, {
+				// Stop at second breakpoint and set a1.
+				execute: func() {
+					tester := &helperForSetVariable{t, client}
+
+					handleStop(t, client, 1, "main.barfoo", -1)
+					// Test: set string 'a1' in main.barfoo.
+					// This shouldn't affect 'a1' in main.foobar - we will check that in the next breakpoint.
+					locals := tester.variables(1001)
+					expectVarExact(t, locals, -1, "a1", "a1", `"bur"`, noChildren)
+					tester.expectSetVariable(1001, "a1", `"fur"`)
+					tester.evaluate("a1", `"fur"`, noChildren)
+					// We will check a1 in main.foobar isn't affected from the next breakpoint.
```

Comment:
Instead of advancing to the next breakpoint, you could just load another scope for the calling function. Users can do that in the UI by clicking on different functions in the stack.